src/compiler/checker.ts

```diff
@@ -20255,7 +20293,49 @@ namespace ts {
             return isCallOrNewExpression(node.parent) && node.parent.expression === node;
         }
 
-        function checkPropertyAccessExpressionOrQualifiedName(node: PropertyAccessExpression | QualifiedName, left: Expression | QualifiedName, right: Identifier) {
+        function getPropertyForPrivateName(apparentType: Type, leftType: Type, right: PrivateName, errorNode: Node | undefined): Symbol | undefined {
+            let classWithShadowedPrivateName;
+            let container = getContainingClass(right);
+            while (container) {
+                const symbolTableKey = getPropertyNameForPrivateNameDescription(container.symbol, right.escapedText);
+                if (symbolTableKey) {
+                    const prop = getPropertyOfType(apparentType, symbolTableKey);
+                    if (prop) {
+                        if (classWithShadowedPrivateName) {
+                            if (errorNode) {
+                                error(

```

Comment:
the error checking would be better as a separate checkPropertyAccessForPrivateName, and then the get* logic should be a case in getPropertyOfType (possibly *called* from there if it's long).

the code will end up returning a type, even when shadowed, but it will log an error as well.
note this change is predicated on simplifying the symbol's name to just '#foo' instead of '__Class#foo'.

Reply:
I think this one is covered by our discussion of keeping the name mangling for now